initial snapshot does not define collation info

rrd108 opened the pull request:
This PR is intended to fix https://github.com/cakephp/phinx/issues/2225

The added test case passes, so it seems to be fine for me.

However, thi is not a final version as I was not sure about 2 things.

# Question 1

The current solution actually execute the `migration_snapshot` command and creates and writes a new migration file. My guess is that I could do it without creating the file, but I did not find out how.

# Question 2

The current test alters the `articles` table's `title` field's collation. So I guess I have to alter it back to avoid affecting other tests. Do I have to do it, or the test system will drop the table after each tests?

rrd108 committed:
failing test added

rrd108 committed:
failing test added

rrd108 committed:
collation info is added to the column

rrd108 committed:
collation info is added to the column

rrd108 committed:
delete created file

rrd108 committed:
delete created file

rrd108 committed:
fix test

rrd108 committed:
fix test

rrd108 committed:
skip option if collation is not set

rrd108 committed:
skip option if collation is not set

rrd108 commented:
I see, I have to handle the case when the collation is the default. I will get back to it in a few days.

ndm2 commented:
Please compare against the full snapshot like the other snapshot tests do, ideally use `runSnapshotTest()`, see for example the `testSnapshotWithAutoIdCompatibleSignedPrimaryKeys()` test.

The comparison file would be `tests/comparisons/Migration/test_snapshot_with_non_default_collation.php`. And then also please register that comparison file with `MigrationsTest::migrationsProvider()`, that way it will be tested whether the generated snapshot can be applied without errors.

As can be seen from the failing tests, this would affect diffs too, so encoding defaults and changes would need to be properly tested there too.

Might sound a bit pedantic, but when my recent work on Migrations taught me anything, then that its sloppy testing is one of its worst enemies, and a major source of pain when making changes.

rrd108 committed:
avoid adding default collation - should fix sqlite failing tests

rrd108 committed:
avoid adding default collation - should fix sqlite failing tests

rrd108 committed:
collate handling moved to another function

rrd108 committed:
collate handling moved to another function

rrd108 committed:
get default collation from schema

rrd108 committed:
get default collation from schema

rrd108 committed:
fix stan errors

rrd108 committed:
fix stan errors

rrd108 committed:
adding collations for the diff

rrd108 committed:
adding collations for the diff

rrd108 committed:
adding collation to another diff

rrd108 committed:
adding collation to another diff

rrd108 commented:
I still have to fix what @ndm2 suggested.

amayer5125 committed:
Fix Phinx Feature Flags Link Syntax

markstory committed:
Merge pull request #646 from amayer5125/fix-feature-flags-link

Fix Phinx Feature Flags Link Syntax

rrd108 committed:
use snapshot file

rrd108 committed:
use snapshot file

rrd108 commented:
Do I have to do anything more on this?

LordSimal reviewed (approved):
LGTM

ndm2 commented on tests/comparisons/Diff/addRemove/the_diff_add_remove_mysql.php:
```diff
@@ -49,6 +50,7 @@ public function down(): void
         $this->table('articles')
             ->addColumn('excerpt', 'text', [
                 'after' => 'title',
+                'collation' => 'utf8_general_ci',
```
This is the same as the default collation. This being here seems inconsistent with the snapshot behavior. Same for the other diff. It's also inconsistent with itself when a table is being dropped and then recreated, for example the `tags` table in the `default` diff, when recreated in `down()` it has no `collation` option set.

We should probably also cover the case where the comparison schema (the `.lock` file) has tables with columns using non-default collations.

rrd108 replied:
Is it connected to your later comment https://github.com/cakephp/migrations/pull/645#issuecomment-1755534423 ?

Or is it something what I should fix somehow?

rrd108 replied:
@ndm2  I am waiting for your reply

ndm2 replied:
> Is it connected to your later comment [#645 (comment)](https://github.com/cakephp/migrations/pull/645#issuecomment-1755534423) ?

No, that was just about none of the snapshot apply testing working. I've pushed [a fix](https://github.com/cakephp/migrations/pull/648) for it. This PR will need to be rebased accordingly.

> Or is it something what I should fix somehow?

I think so. You've made changes to specifically handle the case where the collation matches the default collation, but this doesn't seem to be applied consistently, as the `collation` option is present here, despite that being the table's default collation.

rrd108 replied:
Ok. They do not really have anything with my fix. 
Done in fde1fb325a7f5aea1146f91464cf581bfbd53575

ndm2 commented on tests/TestCase/Command/BakeMigrationSnapshotCommandTest.php:
```diff
@@ -311,4 +311,20 @@ public function assertCorrectSnapshot($bakeName, $result)
+        $connection = ConnectionManager::get('test');
+        assert($connection instanceof Connection);
+
+        $connection->execute('ALTER TABLE articles MODIFY title VARCHAR(255) CHARACTER SET utf8 COLLATE utf8mb3_hungarian_ci');
```
This should be reset afterwards. Verify it by temporarily placing this test before the other snapshot tests. Also note that it removes the comment from the column.

rrd108 replied:
If I have to reset afterwards, I would just create a new table if you are ok with it. So it will have no effect on any other tests.

ndm2 reviewed (changes requested):
Please also add the new snapshot to `\Migrations\Test\TestCase\MigrationsTest::migrationsProvider()`.

ndm2 commented:
So I just realized that `\Migrations\Test\TestCase\MigrationsTest::testMigrateSnapshots()` is actually still broken and not testing properly. That would also affect this snapshot right here, which as a foreign key for a non-existent table.

I'll try to get that fixed somewhat quickly.

rrd108 committed:
new snapshot added to migrationsProvider

rrd108 committed:
new snapshot added to migrationsProvider

rrd108 commented:
> Please also add the new snapshot to `\Migrations\Test\TestCase\MigrationsTest::migrationsProvider()`.

added in 9052c29

ndm2 committed:
Fix snapshot migration tests.

Only the first snapshot was being tested, as subsequent calls to
`migrate()` will not re-read the migration files from disk, but use
cached instances from the first call.

markstory committed:
Merge pull request #648 from cakephp/4.x-fix-snapshot-tests

4.x - Fix snapshot migration tests.

rrd108 committed:
collation reset

rrd108 committed:
collation reset

rrd108 committed:
get previous test pass after rebase

rrd108 committed:
remove default collation from diff files

rrd108 committed:
Merge remote-tracking branch 'origin/missingSnapshotCollation' into missingSnapshotCollation

rrd108 committed:
fix php-cs errors

rrd108 commented:
@ndm2 As I see tha failing tests coming from other test, not related to this one.

ndm2 commented:
I'm note sure what exactly you did, but it looks like rebasing somehow went wrong, all the changes here from existing commits should not be seen as new. The test failures might be a result of that.

rrd108 committed:
newline removed

rrd108 commented:
> I'm note sure what exactly you did, but it looks like rebasing somehow went wrong, all the changes here from existing commits should not be seen as new. The test failures might be a result of that.

I do not know what did I do wrong.

Anyway I would need some suggestion how to move forward.

I cloned cakephp/migrations to my dev machine and executed `vendor/bin/phpunit tests` without any changes and I get `Tests: 122, Assertions: 396, Errors: 8, Warnings: 1, Skipped: 10`

So it seems that is the starting point for now. 

Re-adding my changes to this newly cloned repo does not change the number of errors, so I guess my additions are ok, but the checks will certainly fail.

rrd108 committed:
newline removed

rrd108 commented:
anyway it seems no error on the pipeline.

dereuromark commented:
I wonder if this could be squash merged 
But it would be safer to squash and rebase Prior to merging
